File: src/dtw/interpolate.ts
import { TimeSeriesFrame, Landmark } from './types';

/**
 * TODO: Interpolação temporal para uniformizar a quantidade de frames.
 * - Reamostragem para N frames
 * - Interpolação linear entre frames
 */
export function interpolateToLength(frames: TimeSeriesFrame[], targetLength: number): TimeSeriesFrame[] {
  if (frames.length === 0 || targetLength <= 0) return [];
  if (frames.length === targetLength) return frames;

  const result: TimeSeriesFrame[] = [];
  const ratio = (frames.length - 1) / (targetLength - 1);

  for (let i = 0; i < targetLength; i++) {
    const pos = i * ratio;
    const low = Math.floor(pos);
    const high = Math.min(Math.ceil(pos), frames.length - 1);
    const t = pos - low;

    if (low === high) {
      result.push(frames[low]);
      continue;
    }

    const interpolated: Landmark[] = frames[low].landmarks.map((lm, idx) => {
      const other = frames[high].landmarks[idx];
      return {
        x: lm.x + t * (other.x - lm.x),
        y: lm.y + t * (other.y - lm.y),
        z: lm.z + t * (other.z - lm.z),
        visibility: lm.visibility,
      };
    });

    result.push({
      timestamp: frames[low].timestamp + t * (frames[high].timestamp - frames[low].timestamp),
      landmarks: interpolated,
    });
  }

  return result;
}

File: dtw/interpolate.py
import math
from .types import Landmark, TimeSeriesFrame


def interpolate_to_length(
    frames: list[TimeSeriesFrame], target_length: int
) -> list[TimeSeriesFrame]:
    if not frames or target_length <= 0:
        return []
    if len(frames) == target_length:
        return frames

    result: list[TimeSeriesFrame] = []
    ratio = (len(frames) - 1) / (target_length - 1)

    for i in range(target_length):
        pos = i * ratio
        low = math.floor(pos)
        high = min(math.ceil(pos), len(frames) - 1)
        t = pos - low

        if low == high:
            result.append(frames[low])
            continue

        interpolated = [
            Landmark(
                x=a.x + t * (b.x - a.x),
                y=a.y + t * (b.y - a.y),
                z=a.z + t * (b.z - a.z),
                visibility=a.visibility,
            )
            for a, b in zip(frames[low].landmarks, frames[high].landmarks)
        ]

        timestamp = frames[low].timestamp + t * (
            frames[high].timestamp - frames[low].timestamp
        )
        result.append(TimeSeriesFrame(timestamp=timestamp, landmarks=interpolated))

    return result
